rcon/steam_utils.py

```diff
@@ -2,35 +2,72 @@
 import logging
 import math
 import time
+from dataclasses import dataclass
 from functools import wraps
-from typing import Any, Mapping
+from typing import Any, Iterable, Sequence
 
 import steam.exceptions
+from sqlalchemy import or_, select
+from sqlalchemy.dialects.postgresql import JSONB
+from sqlalchemy.orm import Session, joinedload
 from steam.webapi import WebAPI
-from dataclasses import dataclass
 
 from rcon.cache_utils import ttl_cache
-from rcon.models import PlayerSteamID, SteamInfo
-from rcon.types import SteamBanResultType, SteamBansType
+from rcon.models import PlayerSteamID, SteamInfo, enter_session
+from rcon.types import SteamBansType, SteamInfoType, SteamPlayerSummaryType
 from rcon.user_config.steam import SteamUserConfig
+from rcon.utils import batched
 
 logger = logging.getLogger(__name__)
-
 last_steam_api_key_warning = datetime.datetime.now()
 
+STEAM_API: WebAPI | None = None
+
+
+def get_steam_api() -> WebAPI:
+    """Maintain a single initialized instance of the Steam WebAPI"""
+    global STEAM_API
+    if STEAM_API is None:
+        api_key = get_steam_api_key()
+        STEAM_API = WebAPI(key=api_key)
+
+    return STEAM_API
+
+
+def get_steam_api_key() -> str | None:
+    """Return the servers steam API key if configured"""
+    global last_steam_api_key_warning
+
+    config = SteamUserConfig.load_from_db()
+    if config.api_key is None or config.api_key == "":
+        timestamp = datetime.datetime.now()
+        # Only log once every 5 minutes or it is super spammy
+        if (timestamp - last_steam_api_key_warning).total_seconds() > 300:
+            logger.warning("Steam API key is not set some features will be disabled.")
```

Comment:
do we need this at all? What about checking this at startup once and then never log it again? :D

Reply:
The steam api key can be set (or invalidated) at any point via the config page now, not just at startup.
This is also the same check that exists in current code.

Reply:
Maybe we just bump up the time frame it logs for? I added it because it was spamming my logs while I was testing other stuff in other branches and found it quite infuriating.

I think caranci is right though since it can be set while CRCON is live, it's still useful,especially if we end up in a situation where we're accidentally resetting user config values again (although I think we fixed all of those)

Reply:
Hmm, in this case, I would either up the timer to an incredible high value (like one hour, or maybe even one day) or omit this log message at all. Configuring (or better, not doing so) a steam API key is an explicit action, I don't think why we would need that log message at all.

Reply:
I think the value comes from users who might not be aware they need to input a steam API key for them to take advantage of the feature but will spot the warning in the logs, I think an hour is fine though, it used to output constantly and I just wanted to cut down on that because it was drowning out everything else.